src/client/components/common/toast/toast.reducer.js

```diff
@@ -0,0 +1,68 @@
+import _omit from 'lodash/omit';
+
+import {
+  ADD_MESSAGE,
+  REMOVE_MESSAGE,
+} from './toast.actions';
+
+const defaultState = {
+  messages: {},
+  ids: [],
+};
+
+const messageHash = (message) => {
+  const title = message.title || '';
+  const text = JSON.stringify(message.text);
+  return `${message.type}-${title}-${text}-${message.isHTML}`;
+};
+
+const uniqMessages = (stateMessages, message) => {
+  const newState = {};
+  Object.keys(stateMessages).forEach((id) => {
+    if (messageHash(stateMessages[id]) !== messageHash(message)) {
+      newState[id] = stateMessages[id];
+    }
+  });
+
+  return newState;
+};
+
+const toast = (state = defaultState, action) => {
+  let id;
+  let messages;
+  switch (action.type) {
+    case ADD_MESSAGE:
```

Comment:
The logic looks unnecessary complicated for me

1. I think `uniqMessages` (+ `messageHash`) method isn't useful here (at least for now) and only makes things harder to understand, as message already has 'id' field which for sure unique by itself

2. `ids` field is not needed in state as it only duplicates messages data and adding additional field to maintain. It is already led to some weird checks like `...state.ids.filter(messageId => messages[messageId])` (when adding message)

Reply:
1. This was done in order to prevent the addition of identical error messages to the store.

Reply:
Anyway, as for me – I'd prefer messages duplication rather than complicated code that's not clear at the first glance. This file can be refactored to the small amount of lines in this case

Also messages duplication can be intended and needed by design (for example I'd prefer showing 'Unexpected error occurred' error shown multiple times (or re-showing one) rather than no response from UI when clicking the same button) And it's seems not possible with current code without adding some other additional parameter

I'd rather stick with duplicated messages by default and setting optional 'id' upfront for showing the messages that need to be unique.